dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/sharing/DashboardCascadeSharingTest.java

```diff
@@ -0,0 +1,565 @@
+import org.hisp.dhis.schema.descriptors.ProgramStageSchemaDescriptor;
+import org.hisp.dhis.schema.descriptors.TrackedEntityAttributeSchemaDescriptor;
+import org.hisp.dhis.schema.descriptors.VisualizationSchemaDescriptor;
+import org.hisp.dhis.security.acl.AccessStringHelper;
+import org.hisp.dhis.security.acl.AclService;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.hisp.dhis.trackedentity.TrackedEntityAttributeDimension;
+import org.hisp.dhis.trackedentity.TrackedEntityDataElementDimension;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserGroup;
+import org.hisp.dhis.user.UserService;
+import org.hisp.dhis.user.sharing.Sharing;
+import org.hisp.dhis.user.sharing.UserAccess;
+import org.hisp.dhis.user.sharing.UserGroupAccess;
+import org.hisp.dhis.visualization.Visualization;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class DashboardCascadeSharingTest
+    extends CascadeSharingTest
+{
+    @Autowired
+    private UserService _userService;
+
+    @Autowired
+    private AclService aclService;
+
+    @Autowired
+    private CascadeSharingService cascadeSharingService;
+
+    @Autowired
+    private IdentifiableObjectManager objectManager;
+
+    private UserGroup userGroupA;
+
+    private User userA;
+
+    private User userB;
+
+    private Sharing sharingReadForUserA;
+
+    private Sharing sharingReadForUserAB;
+
+    private Sharing sharingReadWriteForUserB;
+
+    private Sharing sharingUserGroupA;
+
+    private Program programA;
+
+    @Override
+    public void setUpTest()
+    {
+        userService = _userService;
+        userGroupA = createUserGroup( 'A', Collections.EMPTY_SET );
+        objectManager.save( userGroupA );
+
+        userA = createUser( 'A' );
+        userA.getGroups().add( userGroupA );
+        userService.addUser( userA );
+
+        userB = createUser( 'B' );
+        userService.addUser( userB );
+
+        sharingReadForUserA = new Sharing();
+        sharingReadForUserA.setPublicAccess( AccessStringHelper.DEFAULT );
+        sharingReadForUserA.addUserAccess( new UserAccess( userA, AccessStringHelper.READ ) );
+
+        sharingReadWriteForUserB = new Sharing();
+        sharingReadWriteForUserB.setPublicAccess( AccessStringHelper.DEFAULT );
+        sharingReadWriteForUserB.addUserAccess( new UserAccess( userB, AccessStringHelper.READ_WRITE ) );
+
+        sharingReadForUserAB = new Sharing();
+        sharingReadForUserAB.setPublicAccess( AccessStringHelper.DEFAULT );
+        sharingReadForUserAB.addUserAccess( new UserAccess( userA, AccessStringHelper.READ ) );
+        sharingReadForUserAB.addUserAccess( new UserAccess( userB, AccessStringHelper.READ ) );
+
+        sharingUserGroupA = new Sharing();
+        sharingUserGroupA.setPublicAccess( AccessStringHelper.DEFAULT );
+        sharingUserGroupA.addUserGroupAccess( new UserGroupAccess( userGroupA, AccessStringHelper.READ ) );
```

Comment:
Such repetitive setup could be shortened to something like below with help of some additional constructors with varargs.  
```java
sharingReadForUserA = new Sharing( DEFAULT, new UserAccess( userA, READ ));
sharingReadWriteForUserB = new Sharing( DEFAULT, new UserAccess( userB, READ_WRITE ));
sharingReadForUserAB = new Sharing( DEFAULT, new UserAccess( userA, AccessStringHelper.READ ), new UserAccess( userB, AccessStringHelper.READ ) );
sharingUserGroupA = new Sharing( DEFAULT, new UserGroupAccess( userGroupA, AccessStringHelper.READ ));
```

Reply:
fixed